### PR TITLE
Optimize lambda copy to improve direct call performance.

### PR DESCRIPTION
**Pull request by MissiontoMars**
## Why are these changes needed?
There exists extra overhead on the task handling code path, which influences direct call performance, such as:
- function object copy
- variable copy in lambda capture, now move can be used in C++14
- C++ move semantic invalid in `MessageWrapper` constructor
- useless code in `ComputeResources`

In Ant's internal java performance testing, these changes improve performance by 10%-20%. The test case code also checked in here.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Can you fix failing tests first?

**Comment by MissiontoMars** [comment]
> Can you fix failing tests first?

OK

File: java/performance_test/src/main/java/io/ray/performancetest/JobInfo.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package io.ray.performancetest;
+
+import com.google.common.base.Preconditions;
+import io.ray.api.call.BaseActorCreator;
+import io.ray.api.runtimecontext.NodeInfo;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JobInfo {
```
Comment (kfstorm) [inline java/performance_test/src/main/java/io/ray/performancetest/JobInfo.java:10]: Do we need this class in the community version?

File: java/performance_test/src/main/java/io/ray/performancetest/Receiver.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package io.ray.performancetest;
+
+import java.nio.ByteBuffer;
+
+public class Receiver {
+
+  private static final String RECEIVER_RECEIVE_QPS = "receiver.receive.qps";
```
Comment (kfstorm) [inline java/performance_test/src/main/java/io/ray/performancetest/Receiver.java:7]: not used

File: java/performance_test/src/main/java/io/ray/performancetest/Source.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Source {
+
+  private static final String SOURCE_SEND_QPS = "source.send.qps";
+  private static final String SOURCE_GET_RT = "source.get.rt";
```
Comment (kfstorm) [inline java/performance_test/src/main/java/io/ray/performancetest/Source.java:17]: not used

File: java/performance_test/src/main/java/io/ray/performancetest/TestUtils.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+public class TestUtils {
+  private static boolean DEV_MODE = false;
+
+  static {
+    if (null != System.getenv("PERF_TEST_DEV")) {
+      DEV_MODE = true;
+    }
+  }
+
+  public static boolean isDevMode() {
```
Comment (kfstorm) [inline java/performance_test/src/main/java/io/ray/performancetest/TestUtils.java:12]: We don't need this in the community version, right?

File: java/performance_test/src/main/java/io/ray/performancetest/test/ActorPerformanceTestCase1.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package io.ray.performancetest.test;
+
+/** 1对1单向ray call，一收一发，测试引擎本身的吞吐. */
```
Comment (kfstorm) [inline java/performance_test/src/main/java/io/ray/performancetest/test/ActorPerformanceTestCase1.java:3]: English, please.

File: java/performance_test/src/main/resources/ray.conf [hunk 6]
```diff
@@ -0,0 +1,5 @@
+ray {
+  job {
+    num-java-workers-per-process: 10
```
Comment (kfstorm) [inline java/performance_test/src/main/resources/ray.conf:3]: We set `ray.job.num-java-workers-per-process` before `Ray.init()`. This file is unnecessary.

File: src/ray/common/task/task_spec.cc [hunk 7]
```diff
@@ -53,24 +53,23 @@ bool TaskSpecification::PlacementGroupCaptureChildTasks() const {
...
-  auto required_resources = MapFromProtobuf(message_->required_resources());
-  auto required_placement_resources =
-      MapFromProtobuf(message_->required_placement_resources());
-  if (required_placement_resources.empty()) {
-    required_placement_resources = required_resources;
-  }
+  auto &required_resources = message_->required_resources();
 
   if (required_resources.empty()) {
-    // A static nil object is used here to avoid allocating the empty object every time.
```
Comment (kfstorm) [inline src/ray/common/task/task_spec.cc:64]: Why remove this comment?

File: src/ray/common/task/task_spec.h [hunk 8]
```diff
@@ -121,15 +126,16 @@ class TaskSpecification : public MessageWrapper<rpc::TaskSpec> {
   /// Return the ObjectIDs of any dependencies passed by reference to this
   /// task. This is recomputed each time, so it can be used if the task spec is
   /// mutated.
-  ///
+  /// \param add_dummy_dependency whether add dummy object to dependencies or not.
```
Comment (kfstorm) [inline src/ray/common/task/task_spec.h:129]:
```suggestion
  /// \param add_dummy_dependency whether to add a dummy object in the returned objects.
```

Comment (kfstorm) [reply src/ray/common/task/task_spec.h:129]: It seems that the comment is added on the wrong method. It should be added to `GetDependencies`.

File: src/ray/core_worker/core_worker.cc [hunk 9]
```diff
@@ -2208,11 +2208,11 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
   // execution service.
   if (request.task_spec().type() == TaskType::ACTOR_TASK) {
     task_execution_service_.post(
-        [=] {
+        [this, request, reply, callback = std::move(send_reply_callback)] {
```
Comment (kfstorm) [inline src/ray/core_worker/core_worker.cc:2211]:
```suggestion
        [this, request, reply, send_reply_callback = std::move(send_reply_callback)] {
```

This is much clear to me.

**Comment by kfstorm** [comment]
BTW, it would be great if we can actually run the Java performance tests in UT for a while (e.g. 1min).

File: java/performance_test/src/main/java/io/ray/performancetest/JobInfo.java (hunk 1 above)
Comment (MissiontoMars) [reply java/performance_test/src/main/java/io/ray/performancetest/JobInfo.java]: removed.

File: java/performance_test/src/main/java/io/ray/performancetest/Receiver.java (hunk 2 above)
Comment (MissiontoMars) [reply java/performance_test/src/main/java/io/ray/performancetest/Receiver.java]: removed.

File: java/performance_test/src/main/java/io/ray/performancetest/Source.java (hunk 3 above)
Comment (MissiontoMars) [reply java/performance_test/src/main/java/io/ray/performancetest/Source.java]: removed.

File: java/performance_test/src/main/java/io/ray/performancetest/TestUtils.java (hunk 4 above)
Comment (MissiontoMars) [reply java/performance_test/src/main/java/io/ray/performancetest/TestUtils.java]: removed.

**Comment by rkooo567** [comment]
Btw, please remove the tag once the PR is ready :)!

**Review by kfstorm — APPROVED** [review]
LGTM

**Comment by rkooo567** [comment]
There are a lot of test failures including Windows one. Can you merge the latest master?

**Comment by kfstorm** [comment]
test_placement_group and MAC_WHEELS are flakey on master.
Build Images in buildkite failed with "AssertionError: Found (0) matches for '*cp36*-manylinux*' instead of 1" should be unrelated because this PR has nothing to do with Python code.

Merging.

**Pull request merged**